tests/bsim/bluetooth/host/att/sequential/dut/src/main.c

```diff
@@ -0,0 +1,368 @@
+};
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	struct bt_le_conn_param *param;
+	struct bt_conn *conn;
+	int err;
+
+	err = bt_le_scan_stop();
+	if (err) {
+		FAIL("Stop LE scan failed (err %d)", err);
+		return;
+	}
+
+	char str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, str, sizeof(str));
+
+	LOG_DBG("Connecting to %s", str);
+
+	param = BT_LE_CONN_PARAM_DEFAULT;
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, param, &conn);
+	if (err) {
+		FAIL("Create conn failed (err %d)", err);
+		return;
+	}
+}
+
+static void connect(void)
+{
+	struct bt_le_scan_param scan_param = {
+		.type = BT_LE_SCAN_TYPE_ACTIVE,
+		.options = BT_LE_SCAN_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	UNSET_FLAG(is_connected);
+
+	int err = bt_le_scan_start(&scan_param, device_found);
+
+	ASSERT(!err, "Scanning failed to start (err %d)\n", err);
+
+	LOG_DBG("Central initiating connection...");
+	WAIT_FOR_FLAG(is_connected);
+	LOG_INF("Connected as central");
+
+	/* No security support on the tinyhost unfortunately */
+}
+
+static ssize_t written_to(struct bt_conn *conn,
+			  const struct bt_gatt_attr *attr,
+			  const void *buf,
+			  uint16_t len,
+			  uint16_t offset,
+			  uint8_t flags)
+{
+	static int nwrites;
+
+	LOG_INF("written to: handle 0x%x len %d flags 0x%x",
+		attr->handle,
+		len,
+		flags);
+
+	LOG_HEXDUMP_DBG(buf, len, "Write data");
+
+	if (nwrites == 0) {
+		/* Suspend on the first write, which is an ATT Request */
+		LOG_INF("suspending HCI TX thread");
+		k_thread_suspend(bt_testing_tx_tid_get());
+	}
+
+	if (++nwrites == 3) {
+		SET_FLAG(three_writes);
+	}
+
+	return len;
+}
+
+#define test_service_uuid                                                                          \
+	BT_UUID_DECLARE_128(BT_UUID_128_ENCODE(0xf0debc9a, 0x7856, 0x3412, 0x7856, 0x341278563412))
+#define test_characteristic_uuid                                                                   \
+	BT_UUID_DECLARE_128(BT_UUID_128_ENCODE(0xf2debc9a, 0x7856, 0x3412, 0x7856, 0x341278563412))
+
+BT_GATT_SERVICE_DEFINE(test_gatt_service,
+		       BT_GATT_PRIMARY_SERVICE(test_service_uuid),
+		       BT_GATT_CHARACTERISTIC(test_characteristic_uuid,
+					      (BT_GATT_CHRC_READ |
+					       BT_GATT_CHRC_WRITE |
+					       BT_GATT_CHRC_NOTIFY |
+					       BT_GATT_CHRC_INDICATE),
+					      BT_GATT_PERM_READ | BT_GATT_PERM_WRITE,
+					      /* TODO: callbacks */
+					      NULL, written_to, NULL),
+		       BT_GATT_CCC(NULL,
+				   BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+		       );
+
+static uint8_t notified(struct bt_conn *conn,
+			struct bt_gatt_subscribe_params *params,
+			const void *data,
+			uint16_t length)
+{
+	static uint8_t notification[] = "notification";
+	static uint8_t indication[] = "indication";
+	static int num_nfy;
+
+	ASSERT(length >= sizeof(indication), "Unexpected data");
```

Comment:
I would suggest to use `if` and `FAIL` instead of ASSERT here, as ASSERT can be disabled by a Kconfig, and I guess you always want to fail in this case? 

This applies to all `ASSERT` here

Reply:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/68428270/590926cd-06b6-41a6-b588-41815c12a284)

It's defined in `common/utils.h` and is not disableable.

Reply:
Ah, another one of these :D 

Had another PR where I had this exact conversation, and suggested to rename the locally defined ASSERT to something like `BSIM_ASSERT` to avoid this confusion :D 

At some point we probably want to create a testing framework for BSIM to have a set of common functionality like this across all BSIM tests

Reply:
100% agree for the framework